src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Pkcs10CertificationRequestInfo.cs

```diff
@@ -36,52 +37,46 @@ internal class Pkcs10CertificationRequestInfo
             }
         }
 
-        private byte[] Encode()
-        {
-            // CertificationRequestInfo::= SEQUENCE {
-            //   version INTEGER { v1(0) } (v1,...),
-            //   subject Name,
-            //   subjectPKInfo SubjectPublicKeyInfo{ { PKInfoAlgorithms } },
-            //   attributes[0] Attributes{ { CRIAttributes } }
-            // }
-            //
-            // Attributes { ATTRIBUTE:IOSet } ::= SET OF Attribute{{ IOSet }}
-
-            byte[][] attrSet = Attributes.SegmentedEncodeAttributeSet();
-
-            // Replace the tag with ContextSpecific0.
-            attrSet[0][0] = DerSequenceReader.ContextSpecificConstructedTag0;
-
-            return DerEncoder.ConstructSequence(
-                s_encodedVersion,
-                Subject.RawData.WrapAsSegmentedForSequence(),
-                PublicKey.SegmentedEncodeSubjectPublicKeyInfo(),
-                attrSet);
-        }
-
         internal byte[] ToPkcs10Request(X509SignatureGenerator signatureGenerator, HashAlgorithmName hashAlgorithm)
         {
             // State validation should be runtime checks if/when this becomes public API
             Debug.Assert(signatureGenerator != null);
             Debug.Assert(Subject != null);
             Debug.Assert(PublicKey != null);
 
-            // CertificationRequest ::= SEQUENCE {
-            //   certificationRequestInfo CertificationRequestInfo,
-            //   signatureAlgorithm AlgorithmIdentifier{ { SignatureAlgorithms } },
-            //   signature BIT STRING
-            //  }
-
-            byte[] encoded = Encode();
-            byte[] signature = signatureGenerator.SignData(encoded, hashAlgorithm);
             byte[] signatureAlgorithm = signatureGenerator.GetSignatureAlgorithmIdentifier(hashAlgorithm);
+            AlgorithmIdentifierAsn signatureAlgorithmAsn;
+
+            // Deserialization also does validation of the value.
+            signatureAlgorithmAsn = AsnSerializer.Deserialize<AlgorithmIdentifierAsn>(signatureAlgorithm, AsnEncodingRules.DER);
 
-            EncodingHelpers.ValidateSignatureAlgorithm(signatureAlgorithm);
+            SubjectPublicKeyInfoAsn spki = new SubjectPublicKeyInfoAsn();
+            spki.Algorithm = new AlgorithmIdentifierAsn { Algorithm = PublicKey.Oid, Parameters = PublicKey.EncodedParameters.RawData };
+            spki.SubjectPublicKey = PublicKey.EncodedKeyValue.RawData;
+
+            CertificationRequestInfoAsn requestInfo = new CertificationRequestInfoAsn
+            {
+                Version = 0,
+                Subject = this.Subject.RawData,
+                SubjectPublicKeyInfo = spki,
+                Attributes = Attributes.Select(e => new X501AttributeAsn(e)).ToArray()
```

Comment:
I notice that all of your multi-line object initializers omit the optional trailing comma.  We don't have a style rule one way or another, but my personal preference is to always write trailing commas (when multiline), because it cuts down on noise in future edits.

Reply:
Fine, I will add the commas.